Support signing a pre-calculated hash

jls5177 opened the pull request:
This code change adds a new function that takes a pre-computed hash of a message that should be signed (as apposed to a message). This change allows one process to generate the hash of message and another process (over RPC in my case) to perform the signing.

sybrenstuvel commented on rsa/pkcs1.py:
```diff
@@ -283,6 +279,32 @@ def sign(message, priv_key, hash):
+    """
+    
+    # Verify hash_method is a valid hash algorithm
+    if hash_method not in HASH_ASN1:
```
Why perform this check here? It's the same as in `sign_hash()` anyway.

jls5177 replied:
This was an oversight in my original commit. This is removed in the new commit.

jls5177 commented:
I added 2 new unit tests to cover the new hashing flow and updated the usage docs with the new flow. I am new to this type of documentation. Let me know if I missed anything.

sybrenstuvel commented on rsa/pkcs1.py:
```diff
@@ -283,6 +279,28 @@ def sign(message, priv_key, hash):
+    """
+
+    # Calculate the hash and perform the signing
+    return sign_hash(hash(message, hash_method), priv_key, hash_method)
```
Please splitt his up into two lines, using a helper variable. The way it is now it does too much on one line; a traceback referring to an error in this line will be too ambiguous.

jls5177 replied:
Sure, this will be fixed in the next commit.

sybrenstuvel commented on tests/test_pkcs1.py:
```diff
@@ -106,3 +106,35 @@ def test_multiple_signings(self):
+        signature1 = pkcs1.sign_hash(msg_hash, self.priv, 'SHA-256')
+        print("\tSignature1: %r" % signature1)
+
+        # Calculate the signature using the original method
```
The comment "the original method" implies that the reader knows that your method is something that was added later. This is only true within the context of this pull request, but not later once everything has been merged.

jls5177 replied:
Agreed, I switched it to now say `unified method`.

sybrenstuvel commented on tests/test_pkcs1.py:
```diff
@@ -106,3 +106,35 @@ def test_multiple_signings(self):
+        """Hashing and then signing should match with directly signing the message. """
+
+        message = b'je moeder'
+        print("\tMessage:   %r" % message)
```
Please don't print things in a unit test.

jls5177 replied:
I was using the other test cases in this file as a template and many of them had prints.  The prints are now removed.

sybrenstuvel commented:
Another thing: `hash()` is not a good name for a function, as it's also [a built-in Python function](https://docs.python.org/3/library/functions.html#hash).

sybrenstuvel commented:
Hmm not too fond of "generate", maybe "compute_hash"?

sybrenstuvel commented:
Thanks, it's in!
